### ISSUE TITLE
api: [feature] Implement a new encryption client

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio-go#611]
- Asymmetric encryption key (RSA)
- Symmetric encryption key (AES) 

An object saved on the server would have additional headers

```
X-Amz-Meta-X-Amz-Iv: okeTSpk8YgrBJQXik9CBiw==
X-Amz-Meta-X-Amz-Key: Jktv7Y6Q3SX2VtNBNCO5yZARKfiq3yXIG5yPnU0hz9SbSSS9vZToWUdfT89bEJV6
X-Amz-Meta-X-Amz-Matdesc: {}
```

Relevant debug output of reading a file. 
```
mc: <DEBUG> GET /736e7093-f11b-4cbb-b851-e7a4079a09d9-170212-042022/638fa573-0c15-4af9-9e94-0e872ee1ae2e HTTP/1.1
Host: 127.0.0.1:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.3 mc/2017-02-09T21:44:53Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170212/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170212T192910Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 80
Accept-Ranges: bytes
Content-Type: application/octet-stream
Date: Sun, 12 Feb 2017 19:29:10 GMT
Etag: "19abf067dd3a05baa937f56570212f8d"
Last-Modified: Sun, 12 Feb 2017 12:20:23 GMT
Server: Minio/DEVELOPMENT.2017-02-12T01-18-19Z (linux; amd64)
Vary: Origin
X-Amz-Meta-X-Amz-Iv: okeTSpk8YgrBJQXik9CBiw==
X-Amz-Meta-X-Amz-Key: Jktv7Y6Q3SX2VtNBNCO5yZARKfiq3yXIG5yPnU0hz9SbSSS9vZToWUdfT89bEJV6
X-Amz-Meta-X-Amz-Matdesc: {}
X-Amz-Request-Id: 14A2A0E948A578BC

mc: <DEBUG> Response Time:  1.497569ms
```

**Comment by GarimaKapoor** [comment]
@vadmeste do wen have a timeline for this feature. This is a priority. Once you implement in GO other languages an catchup.

**Comment by donatello** [comment]
This is a very useful feature no doubt, but it looks like its a candidate for `mc` than `minio-go` to me. This is so because the encryption scheme is a bit opinionated if implemented inside `minio-go`, not allowing applications to use a different style - for e.g. in some applications, it may be important to encrypt the object names (excluding the bucket names). Additionally there are just two methods being proposed - symmetric and asymmetric - however there are a plethora of choices for the actual encryption ciphers, key lengths, key generation options and bit sizes.

**Comment by vadmeste** [comment]
> This is a very useful feature no doubt, but it looks like its a candidate for mc than minio-go to me. This is so because the encryption scheme is a bit opinionated if implemented inside minio-go, not allowing applications to use a different style - for e.g. in some applications, it may be important to encrypt the object names (excluding the bucket names). Additionally there are just two methods being proposed - symmetric and asymmetric - however there are a plethora of choices for the actual encryption ciphers, key lengths, key generation options and bit sizes.

It would be easier for developers to encrypt/decrypt data by only using minio-go. Besides, I think we can make a generic API that accepts io.Reader (which streams encrypted data) and information that can be saved in object metadata to hold information that helps us to decrypt already stored data.

**Comment by harshavardhana** [comment]
What should be implemented here is S3 published specification first. To be able decrypt AWS SDK objects. Then we can explore other ciphers.

**Comment by harshavardhana** [comment]
Fixed by #613
